### ISSUE TITLE
question on help...

### ISSUE THREAD

**Issue by ebrady** [spf13/cobra#235]
In older versions of cobra I noticed —help|-h would show up in usage but it doesn't seem to be showing up any longer (including in the example usage output in the README).  The same is true for the "help" subcommand that is added at the beginning of Execute().  They both work but usage skips showing them now.

Anyhow, just checking if this was a choice that was made in how the output should look going forward… ?  Or perhaps it can be made optional so the client can choose to show or not show -h|—help and the help subcommand if they wish?  Example based essentially on the code in your README examples:

/path/src/github.com/spf13/app: ./app -h
A Fast and Flexible Static Site Generator built with
                love by spf13 and friends in Go.
                Complete documentation is available at http://hugo.spf13.com

Usage:
  hugo [flags]
  hugo [command]

Available Commands:
  version     Print the version number of Hugo

Flags:
  -a, --author string         Author name for copyright attribution (default "YOUR NAME")
      --config string         config file (default is $HOME/.cobra.yaml)
  -l, --license licensetext   Name of license for the project (can provide licensetext in config)
  -b, --projectbase string    base project directory eg. github.com/spf13/
      --viper                 Use Viper for configuration (default true)

Use "hugo [command] --help" for more information about a command.
/path/src/github.com/spf13/app: tree
.
├── app
├── cmd
│   ├── root.go
│   └── version.go
└── main.go

1 directory, 4 files
/path/src/github.com/spf13/app:

Another thought/aside might be an optional feature to sort the list of subcommands alphabetically… including the help subcommand (if it was in the list, in the older version of cobra I have it comes up at the end instead of alphabetically).

One more thought: something I've noticed in the past is that that if one uses cobra/pflags/viper and sets up an option in a yaml or other control file as "permanent" setting for myself that the —help outputs default setting doesn't reflect that value as the default for that option for me but instead shows whatever the original tool-based hard coded default was for the option.  Same is, if I recall, true for ENV override settings and such.  It would be nice if —help showed what the current default was based on current config file, env, etc settings when viper is in play… that would be more accurate for the user in question running the CLI (as if they do not provide the option their default will be whatever was in the config file, or in the env if env override was available for that option and was set).  I've done a poor job of hacking this in via an early Find() pass and some override variables to "pre-load" CLI opts/defaults before doing Execute() but it's a total hack… doing this also allows me to kick on things like CLI debug modes and things like a JSON output mode early so that all errors are dumped in JSON format (for instance) or debugging is kicked on (even if other opts are bad/etc, I set this pre-pass up to ignore bad opts and grab what it can use and adjust default for the opts it knows about).

Probably should be a few separate issues… but only the help questions are really what might be deemed an issue unless that is an intentional feature which is fine (but I was curious).  Anyhow, thanks for any thoughts!

Cheers,
Erik

**Comment by bogem** [comment]
1. `help` command is always shown now. Fixed in #394
2. Subcommands are always sorted alphabetically. `help` uses `cmd.Commands()` for showing up the sorted subcommands as you can see in [docs](https://godoc.org/github.com/spf13/cobra#Command.Commands).
3. If I clearly understood your thoughts, I think what it will be very complex and redundant for user to show, whence default setting is used.

**Comment by ebrady** [comment]
Thanks for the first two items, appreciated.  

The 3rd item is, to rephrase, is to always show what value will be used for a given option if that option is *not* given/overridden via the CLI.  Right now that is not what happens.  If an env var or my yaml/etc config file sets a new "permanent" default value for an option (which can be used on the CLI itself) I do not see that this is what will be used by the tool if I don't give that option.  I expect the help output to show me what value/setting will be used if I do not give that optional dashed option on the CLI.

Example: Lets say there is a CLI called 'mytool' and it has an optional "--timeout <secs>" CLI options … and the default is 30 seconds (that's the default coded into the tool and cobra/viper).  This option allows env and/or config file overrides if I want to change that, as well as a CLI option.  If I run 'mytool —help' I see that if I don't give a timeout it'll use the default, 30 seconds.  Good so far, that is correct.  No problem with that.  Now I know I don't like that timeout so I set it to 60 seconds via an ENV override (or via the yaml config file)… a "permanent" setting effectively.  At this point if I do not provide the "--timeout" option what will be used my 'mytool'?  For me, 60 seconds will be used.  If I run the command "mytool --help" what should the default be shown as?  I believe it should be shown as 60 seconds, since that is exactly what I'll get if I do not give the "--timeout" option.  Currently it still shows 30 seconds, which is incorrect.  That is NOT the value I'll get if I do not give the "--timeout" option, I'll get 60 seconds.  It would make more sense if the tool gave 60 seconds in this situation as that would be accurate.

As above I hacked a pre-pass into my copy of the library and my tool so that I run 'Find()' early followed by a ParseFlags followed by a viper SetPFlags() type hack I put in place.  This does an "early" pass on the options and loads any settings into my options in viper (be it ENV, config file or CLI).  So if I have debug turned on, it's loaded up.  If I have a JSON output mode turned on, that's loaded up.  If I feed a totally bogus option in, that is ignored at this point (that's caught during the full option pass later)… and if one of the options is '—help', that'll end up dumping the current values (i.e.: honoring the env or cfg file settings and showing those as what will be used… or, if the user gave a CLI opt that was a valid override of even those, it'll show that).  Here's that tool, I'll muck around with the '--config' option so first you see the true default ('local/ark/cfg') set in the tool, then I'll override that via ENV setting and again run a help… now we see the new value for that field that would be used if I don't give a CLI override (or config file override)… and, lastly, I then use the CLI option itself which takes priority along with '--help' and even that is honored:

```
% ./arksync -h | egrep config
2017/03/27 16:28:31.660145   -C, --config string   tool config dir|file (default "/local/ark/cfg")
% setenv ARKSYNC_CONFIG /path/to/override/cfg
% ./arksync -h | egrep config
2017/03/27 16:28:56.709972   -C, --config string   tool config dir|file (default "/path/to/override/cfg")
% ./arksync -h --config /higher/level/override | egrep config
2017/03/27 16:29:16.782998   -C, --config string   tool config dir|file (default "/higher/level/override")
%
```

You can ignore the timestamp prepended (that's just my output library that I've subbed in)… I can turn that off.  Anyhow, this "optional" pre-pass also allows me to detect things like debugging or other controls very early on so if I'm setting up my tool before I fire up 'Execute()' I can still error correctly, dump debugging if the user has used that option, etc… and if I'm in a JSON vs text output mode for the tool, no problem, again, after that "early" find is done I'll know that and I can dump all messages in either format correctly.

So there's two things being discussed in that 3rd part.  The most important one is probably getting the value set correctly… i.e.: if I do not give that option on the CLI what value will the tool use? (i.e.: honor the config file and env… and I suggest even honoring the CLI if someone gives that option like the above).  The other part is that for any "prep" the tool does before it kicks into the 'Execute()' mode for the subcommand it is nice to be able to do a "pre-pass" of cfg/env/cli opts so they can be honored if set, i.e.: *very* early in the program.  If these aren't done as part of this request I guess my suggestion is to break that into a couple of feature requests for the future.  (?)

**Comment by bogem** [comment]
Sorry for the waiting. The comment was really long, so I found only now time to read it.
Actually, it redundantly complicates cobra. There are many problems (as you described above) what we should solve and code, but this feature will solve not so many problems for users. The elegant solution is just to write in README and/or in `help` of your app which value will be used.

**Comment by bogem** [comment]
Also it changes the API, so we should wait spf13 for accepting these changes. And as I know, [he doesn't really like then flag modifies the behaviour of other flag](https://github.com/spf13/cobra/issues/392#issuecomment-285387086).

**Comment by bogem** [comment]
Ping

**Comment by ebrady** [comment]
Thanks for the followup… meant to get back to you and got distracted with work/etc, apologies.   I am fine with spf13 pondering the need and, if deemed useful, considering a clean way to adjust… but only if desired.  

The point I made above has nothing to do with a flag modifying the behavior of another flag.  What I indicated above and showed in the example above is a *single flag* (—config) and showing that one can have 4 ways to set the value of that single flag:
1.  Built in setting for the option, built into the tool
2. Config file setting to override the built-in setting
3. Environment variable override is also possible (overriding the config file and built-in settings)
4. Command line option direct use will override all of the above

I showed how the help screen could show the true value for that flag (that will be used in the tool) in the help screen... regardless of which of the above the value was coming from.

The point is that a single flag can be set in these 4 different ways.  I ordered these as that is the order of assessment as to what value should be used (highest number will be used if the option is set that way).  If I set a value in my config file and that flag is configured to support a config file setting then that will be the value of the flag.  If I then set an env var that would override the config file and take precedence as the value to use for the flag.  If I give something on the command line that will always be used and override any config file or env settings.

So my point was basically that if one runs 'mytool —help'  that the help should print out the value that would be used by the tool as the default, e.g.: if that option weren't given on the command line for instance.  Today it will only ever print #1 above.  But if I've set a config setting (#2) or if I've set an env setting for that flag (#3) or if I've used that flag on the CLI (#4) then #1 will not be the value used by the tool.  My default setting is not #1 at that point.  If I've set a config setting in my config file… it's my new default for that option basically.  If I run —help to look at the options and settings that will be used… the default for me is what is in that config file, #2, not #1 (what is shown today is #1 though).  If I have an env var set instead (or in addition) then the default for me is #3, the env var setting, not #1.  

Anyhow, the point is that the help screen effectively becomes inaccurate once I've set a config file setting to set myself a new default value for an option (or via env setting)… so why doesn't the help screen show me the default value that will be used for me and my settings?… since that is the value that will be used if I run the tool without the given option on the CLI (as shown with the —config option in the example above, that's what I would like to see where the —help output is being updated as I set env settings or CLI override settings).  I think this part could be done without an API change… and I think it would make the —help screen more accurate/helpful is all

As to "early detection" of options, that might be an "optional" api change… if someone wants to call the package in a pre-pass mode to scan args and leverage any that are valid early that would be possible… but certainly not required.  However, if one wanted to, very early in one's programs startup phase, one could have the lib scan the default opt settings, override them with config file and env settings and CLI settings (per usual ordering and rules for options)… then one could use those settings very early within a program to adjust debug levels or output formats or whatever… during early startup of one's tool.  That might be handy for some tools that have a fair amount of initialization before kicking into executing some subcommand… and wouldn't effect an API change for those using the mechanism as it's designed today (since they wouldn't use the optional pre-pass).  But if it was helpful for a tool then it could be used.  This would be secondary to the 1st part though… just trying to get the help screen to correspond to the value that will be used for an option when I run the tool.
